src/sema/contracts.rs

```diff
@@ -694,6 +698,31 @@ fn check_inheritance(contract_no: usize, ns: &mut ast::Namespace) {
     ns.diagnostics.extend(diagnostics);
 }
 
+/// A contract on substrate requires at least one public message
+fn substrate_requires_public_functions(contract_no: usize, ns: &mut ast::Namespace) {
+    let contract = &mut ns.contracts[contract_no];
+
+    if ns.target.is_substrate()
+        && !ns.diagnostics.any_errors()
+        && contract.is_concrete()
+        && !contract.all_functions.keys().any(|func_no| {
+            let func = &ns.functions[*func_no];
+
+            match func.ty {
+                pt::FunctionTy::Function => func.is_public(),
+                pt::FunctionTy::Fallback | pt::FunctionTy::Receive => true,
+                _ => false,
+            }
+        })
+    {
+        let message = format!("contracts without public storage or functions are not allowed on Substrate, considering declaring this contract 'abstract contract {}'", contract.name);
```

Comment:
```suggestion
        let message = format!("contracts without public storage or functions are not allowed on Substrate. Consider declaring this contract 'abstract contract {}'", contract.name);
```
I think you meant the imperative form here.